ql/ql/src/codeql_ql/ast/internal/Builtins.qll

```diff
@@ -94,3 +94,148 @@ class FloatClass extends PrimitiveType {
+   *  module EquivalenceRelation<T MyT, EdgeSig<MyT>::edgeSig/2 edge> { // the `edge` parameter is not modeled
+   *    class EquivalenceClass;
+   *    EquivalenceClass getEquivalenceClass(MyT a);
+   *  }
+   *}
+   */
+  module EquivalenceRelation {
+    class SigClass extends MockClass::Range {
+      SigClass() { this = "Mock: QlBuiltins::T" }
+
+      override string getName() { result = "T" }
+    }
+
+    /** A mock TypeExpr with classname `T`. */
+    class DummyTTypeExpr extends MockTypeExpr::Range {
+      DummyTTypeExpr() { this = "Mock: QlBuiltins::T" }
+
+      override string getClassName() { result = "T" }
+    }
+
+    module EdgeSig {
+      class EdgeSigModule extends MockModule::Range {
+        EdgeSigModule() { this = "Mock: QlBuiltins::EdgeSig" }
+
+        override string getName() { result = "EdgeSig" }
+
+        override predicate hasTypeParam(int i, string type, string name) {
+          i = 0 and name = "MyT" and type instanceof DummyTTypeExpr
+        }
+
+        override string getMember(int i) { i = 0 and result instanceof EdgeSigPred }
+      }
+
+      class EdgeSigPred extends MockClasslessPredicate::Range {
+        EdgeSigPred() { this = "Mock: QlBuiltins::EdgeSig::edgeSig" }
+
+        override string getName() { result = "edgeSig" }
+
+        override EdgeSigPredParam getParameter(int i) {
+          i = 0 and
+          result.getName() = "a"
+          or
+          i = 1 and
+          result.getName() = "b"
+        }
+      }
+
+      class EdgeSigPredParam extends MockVarDecl::Range {
+        string name;
+
+        EdgeSigPredParam() {
+          this = "Mock: QlBuiltins::EdgeSig::edgeSig::" + name and name = ["a", "b"]
+        }
+
+        override string getName() { result = name }
+
+        override MockTypeExpr::Range getType() { result instanceof DummyTTypeExpr }
+      }
+    }
+
+    class EquivalenceRelationModule extends MockModule::Range {
+      EquivalenceRelationModule() { this = "Mock: QlBuiltins::EquivalenceRelation" }
+
+      override string getName() { result = "EquivalenceRelation" }
+
+      override string getMember(int i) {
+        i = 0 and result instanceof EquivalenceClassClass
+        or
+        i = 1 and result instanceof GetEquivalenceClassPredicate
+      }
+
+      override predicate hasTypeParam(int i, string type, string name) {
+        i = 0 and name = "MyT" and type instanceof DummyTTypeExpr
+        or
+        none() // TODO: `EdgeSig<MyT>::edgeSig/2 edge` is not implemented.
+      }
+    }
+
+    class GetEquivalenceClassPredicate extends MockClasslessPredicate::Range {
+      GetEquivalenceClassPredicate() {
+        this = "Mock: QlBuiltins::EquivalenceRelation::getEquivalenceClass"
+      }
+
+      override string getName() { result = "getEquivalenceClass" }
+
+      override MockVarDecl::Range getParameter(int i) {
+        result.(EdgeSig::EdgeSigPredParam).getName() = "a" and // just reusing another mock node that has the right name and type.
```

Comment:
Is this the reason why mocks have non-unique parents? Perhaps we should avoid reusing nodes.

Reply:
> Is this the reason why mocks have non-unique parents? 

Yes.  

Avoiding non-unique parents would require me to build way more AST nodes, which would require more code, which I didn't want.

Reply:
I wonder if less code is worth the risk of not catching non-unique parent bugs in the future. Is it really that much more code?

Reply:
> Is it really that much more code? 

One way to find out, and that is to implement it.   
There were 5 copies of the `T` type, and two copies of a parameter named `a` (that had type `T`).  

I've split it out now such that even mock AST nodes have unique parents.